packages/iterator-utilities/src/lib/filter.ts

```diff
@@ -0,0 +1,42 @@
+import type { IterableResolvable } from './from';
+import { assertFunction } from './shared/_assertFunction';
+import { toIterableIterator } from './toIterableIterator';
+
+/**
+ * Creates an iterable with the elements that pass the test implemented by the provided function.
+ *
+ * @param iterable The iterator to filter.
+ * @param callbackFn A function to execute for each element produced by the iterator. It should return a truthy value to make the element yielded by the iterator helper, and a falsy value otherwise.
+ * @returns An iterator that produces elements from the given iterator that satisfy the specified test.
+ *
+ * @example
+ * ```typescript
+ * import { filter } from '@sapphire/iterator-utilities';
+ *
+ * const iterable = [1, 2, 3, 4, 5];
+ * console.log([...filter(iterable, (value) => value % 2 === 0)]);
+ * // Output: [2, 4]
+ * ```
+ */
+export function filter<const ElementType, const FilteredType extends ElementType>(
+	iterable: IterableResolvable<ElementType>,
+	callbackFn: (element: ElementType, index: number) => element is FilteredType
+): IterableIterator<FilteredType>;
+export function filter<const ElementType>(
+	iterable: IterableResolvable<ElementType>,
+	callbackFn: (element: ElementType, index: number) => boolean
+): IterableIterator<ElementType>;
+
```

Comment:
In the words of our dragonite, whats with the space

Reply:
![dragonite fuck yo shit up](https://github.com/sapphiredev/utilities/assets/4019718/c1e8f00a-1f59-4d35-b8b5-e86e2ef18714)